File: src/main/java/aunmag/shooter/core/gui/Notification.java
package aunmag.shooter.core.gui;

import aunmag.shooter.core.gui.font.FontStyle;
import aunmag.shooter.core.utilities.Envelope;
import aunmag.shooter.core.utilities.TimeFlow;
import org.joml.Vector4f;

public class Notification extends Component {

    private final Label title;
    private final Label details;
    private final Envelope opacity;
    private final Vector4f color = new Vector4f(1, 1, 1, 0);

    public Notification(TimeFlow time, String title, String details) {
        this.title = new Label(5, 4, 2, 1, title);
        this.details = new Label(5, 5, 2, 1, details, FontStyle.LABEL_LIGHT);
        this.opacity = new Envelope(0.125f, 2.5f, 0.5f, time); // TODO: Try to use tensity, simplify values
        opacity.start(1);
    }

    @Override
    public void render() {
        opacity.update(); // TODO: Move to update

        color.w = opacity.getValue();

        title.setTextColour(color);
        title.render();

        details.setTextColour(color);
        details.render();
    }

    @Override
    protected void onRemove() {
        title.remove();
        details.remove();
        super.onRemove();
    }

    @Override
    public boolean isActive() {
        return super.isActive() && !opacity.isDone();
    }

}

File: src/main/java/aunmag/shooter/game/gui/Notification.kt
package aunmag.shooter.game.gui

import aunmag.shooter.core.gui.Label
import aunmag.shooter.core.gui.font.FontStyle
import aunmag.shooter.core.utilities.Operative
import aunmag.shooter.core.utilities.TimeFlow
import aunmag.shooter.core.utilities.Timer
import aunmag.shooter.core.utilities.UtilsMath
import org.joml.Vector4f

internal class Notification(
        time: TimeFlow,
        title: String,
        details: String
) : Operative() {

    private val timeFadeIn = 0.125f
    private val timeFadeOut = 0.5f
    private val timer = Timer(time, 3.0)
    private val title = Label(5, 4, 2, 1, title)
    private val details = Label(5, 5, 2, 1, details, FontStyle.LABEL_LIGHT)

    init {
        timer.next()
    }

    override fun update() {
        if (timer.isDone) {
            remove()
        }
    }

    override fun render() {
        val timeInitial = timer.target - timer.duration
        val timePassed = timer.time.current - timeInitial
        val timeRemain = timer.target - timer.time.current
        val timeFade = if (timePassed < timeRemain) {
            timePassed / timeFadeIn
        } else {
            timeRemain / timeFadeOut
        }.toFloat()

        val alpha = UtilsMath.limitNumber(timeFade, 0f, 1f)
        val color = Vector4f(1f, 1f, 1f, alpha)

        title.setTextColour(color)
        details.setTextColour(color)

        title.render()
        details.render()
    }

    override fun remove() {
        if (isRemoved) {
            return
        }

        title.delete()
        details.delete()

        super.remove()
    }
}
